list_response.go

```diff
@@ -13,27 +33,51 @@ func newListResponse(resources []interface{}) listResponse {
 	}
 }
 
-// listResponse identifies a query response.
-type listResponse struct {
-	// TotalResults is the total number of results returned by the list or query operation.
-	// The value may be larger than the number of resources returned, such as when returning
-	// a single page of results where multiple pages are available.
-	// REQUIRED
-	TotalResults int
-
-	// ItemsPerPage is the number of resources returned in a list response page.
-	// REQUIRED when partial results are returned due to pagination.
-	ItemsPerPage int
-
-	// StartIndex is a 1-based index of the first result in the current set of the list results.
-	// REQUIRED when partial results are returned due to pagination.
-	StartIndex int
-
-	// Resources is a multi-valued list of complex objects containing the requested resources.
-	// This may be a subset of the full set of resources if pagination is requested.
-	// REQUIRED if TotalResults is non-zero.
-	Resources []interface{}
-}
+type (
+	// ListResponse identifies a paginated query response.
+	ListResponse struct {
```

Comment:
Is there a specific reason to create duplicates here? Is it possible to keep this private otherwise? We should try keeping the public structures to a minimum.

Reply:
It was in order to expose `Resources` as type `[]Resource` as opposed to `[]interface{}`. It also needed to be public so consumers could return it from their callback. I initially made the original one public but I thought the `Resources` or type  `[]interface{}` might make for a confusing interface. Although, we are exposing a helper function for it so perhaps that was an acceptable approach.

Reply:
Good point.

The only addition value that has to be returned is `TotalResults`. I would make a structure that contains `[]Resources` and that total value. (named`Page` for example and keep `listResponse` private)

Reply:
Right we can discern the other params. Good call 👍